plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeSink.java

```diff
@@ -0,0 +1,241 @@
+                }
+                catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                    throw new RuntimeException(e);
+                }
+            });
+            int writableBytes = min(pendingBuffer.writableBytes(), slice.length() - position);
+            pendingBuffer.writeBytes(slice.getBytes(position, writableBytes));
+            position += writableBytes;
+
+            flushIfNeeded(partitionId, false);
+        }
+    }
+
+    private void flushIfNeeded(int partitionId, boolean finished)
+    {
+        SliceOutput buffer = pendingBuffers.get(partitionId);
+        if (!buffer.isWritable() || finished) {
+            if (!buffer.isWritable()) {
+                pendingBuffers.remove(partitionId);
+            }
+            writers.get(partitionId).write(buffer.slice()).whenComplete((result, throwable) -> {
+                if (throwable != null) {
+                    failure.compareAndSet(null, throwable);
+                }
+
+                buffer.reset();
+                freeBuffers.add(buffer);
+
+                CompletableFuture<?> completableFuture;
+                synchronized (this) {
+                    completableFuture = blockedFuture;
+                }
+                if (!completableFuture.isDone()) {
+                    completableFuture.complete(null);
+                }
+            });
+        }
+    }
+
+    @Override
+    public long getMemoryUsage()
+    {
+        if (finished) {
+            return INSTANCE_SIZE;
+        }
+        else {
+            return INSTANCE_SIZE
+                    + (long) writers.size() * INTEGER_INSTANCE_SIZE
+                    + (long) pendingBuffers.size() * INTEGER_INSTANCE_SIZE
+                    + (long) numBuffers * writeBufferSize;
+        }
+    }
+
+    @Override
+    public CompletableFuture<?> finish()
+    {
+        if (finished) {
+            return completedFuture(null);
+        }
+
+        for (Integer partitionId : writers.keySet()) {
+            flushIfNeeded(partitionId, true);
+        }
+        return allOf(writers.values().stream().map(ExchangeStorageWriter::finish).toArray(CompletableFuture[]::new))
+                .thenRun(this::clear)
+                .thenCompose(ignored -> exchangeStorage.createEmptyFile(outputDirectory.resolve(COMMITTED_MARKER_FILE_NAME)))
+                .whenComplete((result, throwable) -> {
+                    if (throwable != null) {
+                        abort();
+                    }
+                    else {
+                        finished = true;
+                    }
+                });
+    }
+
+    @Override
+    public CompletableFuture<?> abort()
+    {
+        if (finished) {
+            return completedFuture(null);
+        }
+        finished = true;
+
+        return allOf(writers.values().stream().map(ExchangeStorageWriter::abort).toArray(CompletableFuture[]::new))
+                .thenRun(this::clear)
```

Comment:
We can't clear first as we need to make sure all outstanding upload futures are done, such that all buffers are added back to `freeBuffers`

Reply:
If the sink is in state `finished` no buffers should be added back to the buffer queue as it will keep retaining them. It feels like the buffers should be cleaned after `finish` is set, as the retained memory tracker will return incorrect memory utilization.